### ISSUE TITLE
Improve & complete URL manipulation support via URITemplate

### ISSUE THREAD

**Issue by Tomboyo** [Tomboyo/lily#31]
As a user, I can easily overwrite components of a URL, overwrite path parameters, and add and remove query parameters so that I can manually create request URLs whenever Lily or the OAS is missing necessary support.

The URITemplate needs to evolve into a fluent API to make adding and overwriting _parts_ of a URL convenient. We don't want to force the user to `URI.create(uriTemplate.toURI().toString() + "?foo=bar")` or perform even more invasive string manipulation when they could do things like `uriTemplate.setQueryParameter("foo", "bar")` or `uriTemplate.setPath("/pets/or/something")` instead. Encoding should be performed automatically so the user doesn't forget a '?' and break their program.

The updated API might look like this:

```java
// https://example.com/pets/{petId = 1234}
var uriTemplate = ...;

// Assume the following statements do not build on one another; each time, the uriTemplate begins as it is above this line. The text {petId = 1234} represents a parameter in the URL which is bound to a variable, and would render as just 1234.

uriTemplate.setPathParameter("petId", 789); // https://example.com/pets/{petId = 789} (null is invalid argument)
uriTemplate.setHost("github.com") // https://github.com/pets/{petId = 1234}
uriTemplate.setScheme("ftp")  // https://example.com/pets/{petId = 1234}
uriTemplate.setPort(8080) // // https://example.com:8080/pets/{petId = 1234}
uriTemplate.setPath("foo/bar/baz") // // https://example.com/foo/bar/baz (no petId anymore!)
uriTemplate.clearQuery() // remove the query string entirely
uriTemplate.setQuery(Map.of("include", List.of("a", "b'))) // // https://example.com/pets/{petId = 1234}?include=a&include=b (order of query parameters may be different)
uriTemplate.setQueryParameter("foo", 1234) // // https://example.com/pets/{petId = 1234}?foo=1234 (can override current parameter value) (null can unset the parameter)
uriTemplate.removeQueryParameter("foo") // would remove foo=1234 if it were part of the URL, and do nothing otherwise. Same as setQueryParameter("foo", null).
```

We should not support a way to set a raw query string ("?foo=bar&...") because it conflicts with `setQueryParameter(String, Object)`: there is no one format that clients will use to encode array and object parameters, so we don't have an obvious way to decode their parameters to a Map internally. As a result, `setQueryParameter` would have to override their query, do nothing, or raise an exception. This doesn't seem like a valuable API given the client has more convenient ways to set parameters one at a time.

We wont support `setTemplate(String)` method that accepts the `scheme://host:port/path/with/{parameters}/` URI template, even without a query parameter, because decoding even that promises to be tricky to do _correctly_. It would be a slightly more convenient API, but setting each part individually is simple and works.

The proposed API will let the user combine URL templating with builders that automatically encode parameters (such as to form explode for the query string).

Now, what about if the user needs to override encoding? For example, path parameters might use the matrix style instead of the default simple style, and not all OAS specifications are good about indicating their preference (and Lily only supports simple at time of writing). We could let the user set a custom encoder (such as an encodings.Encoding function reference), and we could also let the user set a "raw" (already encoded by them) value that that template should _not_ try to encode. Those APIs might look like this:
```
uriTemplate.setRawPathParameter("color", ";color=blue,black,brown"); // https://example.com/;color=blue,black,brown
uriTemplate
  .setPathParameter("petId", List.of("blue", "black", "brown"))
  .setPathEncoder(Encoding::matrix) // https://example.com/;color=blue,black,brown
```

Query parameters do not, as far as I understand, ever get encoded differently than style=form, explode=true (?foo=bar&...). It's legal for clients to use `;` instead of `&`, but that's only a requirement for servers to handle. That said, there is no standard for encoding complex data like lists. These are all valid strategies:
 - `?array[]=foo&array[]=bar`
- `?array=foo&array=bar`
- `?array=foo,bar`
At some point, we could parameterize the form-explode encoder to allow different array and object encoding strategies, which would let the user evolve from using `setRawQueryParameter` to using a customized encoder.

**Comment by Tomboyo** [comment]
It may be a much better strategy to implement (or use an implementation of) [RFC6570](https://www.rfc-editor.org/rfc/rfc6570), like [this](https://github.com/googleapis/google-http-java-client/blob/main/google-http-client/src/main/java/com/google/api/client/http/UriTemplate.java). That gives the client freedom to mix-and-match encoding strategies and freely combine encoded and unencoded characters. Support for level 1 and 4, as in the google implementation, makes sense.

Examples:
```java
UriTemplate.for("https://example.com/foo/{;colors*}/{?name}{&colors}")
  .bind("colors", List.of("red", "green", "blue"))
  .bind("name", "Oscar")
  .expand(); // https://example.com/foo/;colors=red;colors=green;colors=blue/?name=Oscar&colors=red,green,blue
```

We still need a strategy to allow the user to expand according to a customizable strategy (i.e. colors[]=red&colors[]=blue versus colors=red,blue or colors=red&colors=blue), since a lot of different APIs exist which do not conform to RFC6570.

**Comment by Tomboyo** [comment]
On second thought, while RFC6570 is what swagger explicitly supports, we want a tool that lets users work around inconsistent or incorrect swagger specifications. RFC6570-style templates are restrictive, whereas our original idea can easily support arbitrary encodings. This allows lily-http to be useful in domains beyond OpenAPI implementaitons, and ensures we can support custom style-format combinations (including those that may be added in subsequent iterations of OpenAPI).

**Comment by Tomboyo** [comment]
```java
UriTemplate
  .of("https://example.com/{id}/colors/?{colors}&{foo}")
  .bind("id", 123, Encoding::simple)
  .bind("colors", List.of("red", "green", "blue"), Encoding::formExplode)
  .bindRaw("foo", "bar&baz")
  .build();
  // https://example.com/123/colors/?colors=red&colors=green&colors=blue&foo=bar&baz

UriTemplate
  .of("https://example.com/;{colors}/")
  .bind("colors", List.of("red", "green", "blue"), Encoding::matrix)
  .build();
  // https://example.com/;colors=red,green,blue/
```

This seems like a good middle-ground. The template is convenient to write and is extensible. The most obvious problem is that if the user changes from (for example) matrix style to form style, they'll have to remember to also remove the literal ';' from the path. This is where RFC6570 is superior, since it has a strong relationship between the templates and encoding strategies.

I suppose an alternative is to adopt RFC6570's style of adding special character prefixes and suffixes to template strings, then also allow the user to register encoding functions for specific combinations. We could support RFC6570's characters by default, but then the user could add extensions. Like this, perhaps:

```
UriTemplate.customEncoding((prefix, suffix) -> prefix.equals("[]") && suffix.equals("*") ? myEncoding : null);

UriTemplate
  .of("https://exmaple.com/{id}/{?foo*}{&bar}{[]baz}")
  .bind("id", 123)
  .bind("foo", List.of("a", "b"))
  .bind("bar", List.of("c", "d"))
  .bind("baz", List.of("e", "f"))
  .build();
  // https://example.com/123/?foo=a&foo=b&bar=c,d&baz[]=e&baz[]=f
```

That could be the best of both worlds. The MVP implementation can get away without support for extensions, but we have a clear path to support arbitrary encodings.

**Comment by Tomboyo** [comment]
For posterity:

I've adopted the strategy wherein we pass encoding functions into the UriTemplate, allowing us to use RFC6570-style encoders as well as custom encoders. Parameters in the uri can be used for both path parameters and query parameters, and the API doesn't make any distinction between the two, which keeps the API very simple.

We can `withTemplate` and `appendTemplate` to work around incorrect or incomplete templates set by the operation builder. In particular, this gives users a way to set query strings with the current rendition of Lily.

We can override and remove bindings, which lets us again work around undesired behavior from the operation builders. We can also pass in URL-encoded strings directly for those cases where using an encoding function is more work than its worth.

I've also updated the Encoding functions to better match RFC6570. For example, we now differentiate between `form` and `formContinuation`, which clarifies when a leading `?` should be part of the string expansion.
